## Session #1 Homework

### Question 1

What's the version of NumPy that you installed?

You can get the version information using the `__version__` field:

In [1]:
import numpy as np
np.__version__

'1.23.2'

### Getting the data 


For this homework, we'll use the Car price dataset. Download it from 
[here](https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv).

You can do it with wget:

```bash
wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
```

Or just open it with your browser and click "Save as...".

**I saved dataset into the folder data.**

Now read it with Pandas.


In [2]:
import pandas as pd

In [3]:
dataset = pd.read_csv('data/data.csv')
dataset.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

### Question 2

How many records are in the dataset?

In [4]:
print('Records : ', dataset.shape[0])

Records :  11914


**Answer: 11914**

### Question 3

Who are the most popular car manufacturers (top-3) according to the dataset?

In [5]:
dataset.Make.value_counts().head(3)

Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

**Answer: Chevrolet, Ford, Volkswagen**

### Question 4

What's the number of unique Audi car models in the dataset?

In [6]:
len(dataset[dataset['Make'] == 'Audi'].Model.unique())

34

**Answer: 34**

### Question 5

How many columns in the dataset have missing values?

In [7]:
sum(dataset.isna().sum() != 0)

5

**Answer: 5**

### Question 6

* Find the median value of "Engine Cylinders" column in the dataset.
* Next, calculate the most frequent value of the same "Engine Cylinders".
* Use the `fillna` method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
* Now, calculate the median value of "Engine Cylinders" once again.

Has it changed?

> Hint: refer to existing `mode` and `median` functions to complete the task.

In [8]:
# 1
dataset['Engine Cylinders'].median()

6.0

In [9]:
# 2
dataset['Engine Cylinders'].mode()

0    4.0
dtype: float64

In [10]:
# 3
dataset['Engine Cylinders'].fillna(value=dataset['Engine Cylinders'].mode()[0], inplace=True)

In [11]:
# 4
dataset['Engine Cylinders'].median()

6.0

**Answer: No**

### Question 7

* Select all the "Lotus" cars from the dataset.
* Select only columns "Engine HP", "Engine Cylinders".
* Now drop all duplicated rows using `drop_duplicates` method (you should get a dataframe with 9 rows).
* Get the underlying NumPy array. Let's call it `X`.
* Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
* Invert `XTX`.
* Create an array `y` with values `[1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]`.
* Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
* What's the value of the first element of `w`?

> **Note**: You just implemented linear regression. We'll talk about it in the next lesson.

In [12]:
dataset[["Engine HP", "Engine Cylinders"]][dataset.Make == 'Lotus'].drop_duplicates()

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [13]:
X = dataset[["Engine HP", "Engine Cylinders"]][dataset.Make == 'Lotus'].drop_duplicates().to_numpy()

In [14]:
XTX = X.T @ X

In [15]:
XTX_inv = np.linalg.inv(XTX)

In [16]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])

In [17]:
w = XTX_inv @ X.T @ y
w[0]

4.594944810094551

**Answer: 4.594944810094551**